src/clj/fluree/db/indexer/garbage.cljc

```diff
@@ -0,0 +1,91 @@
+            [fluree.db.util.core :as util]
+            [fluree.db.util.log :as log]))
+
+#?(:clj (set! *warn-on-reflection* true))
+
+(defn garbage-meta-map
+  "Garbage file metadata that will be passed to the
+  garbage cleaning function that allows it to report
+  on status and have filename addresses."
+  [{:keys [prev-index garbage] :as idx-root}]
+  (when garbage ;; first index will not have garbage
+    (assoc garbage
+      ;; retain not the current index file, but the previous one
+      ;; where the garbage in this file originated
+      :t (:t prev-index)
+      :index (:address prev-index))))
+
+(defn trace-idx-roots
+  [conn commit]
+  (go-try
+   (loop [next-idx-root (<! (storage/read-db-root conn (-> commit :index :address)))
+          garbage       []]
+     (if (or (nil? next-idx-root) ;; no more idx-roots
+             (util/exception? next-idx-root)) ;; if idx-root already deleted, will be exception
+       garbage
+       (let [garbage-meta  (garbage-meta-map next-idx-root)
+             prev-idx-addr (get-in next-idx-root [:prev-index :address])
+             garbage*      (if garbage-meta
+                             (conj garbage garbage-meta)
+                             garbage)]
+         (recur (<! (storage/read-db-root conn prev-idx-addr))
+                garbage*))))))
+
+(defn clean-garbage-record
+  "Cleans up a complete garbage file, which will contain
+  many index segment garbage items within it. Garbage file
+  looks like:
+  {:alias 'ledger-alias',
+   :branch 'main',
+   :garbage ['fluree:file://.../index/segment/1.json', ...]
+   :t 1}"
+  [conn {:keys [address index t] :as _garbage-map}]
+  (go
+    (let [{:keys [alias branch garbage]} (<! (storage/read-garbage conn address))]
+      (log/info "Removing" (count garbage) "unused index segments (garbage) from ledger"
+                alias "branch" branch "from index-t of" t)
+
+      ;; first delete all index segment files
+      (doseq [garbage-item garbage]
+        ;; note if the file was already deleted there could be an exception.
+        ;; this might happen if the server shutdown in the middle of a garbage
+        ;; exceptions are logged downstream, so just swallow exception here
+        ;; and keep rocessing.
+        (<! (storage/delete-garbage-item conn garbage-item)))
+
+      ;; delete main garbage record
+      (<! (storage/delete-garbage-item conn address))
+
+      ;; then delete the parent index root
+      (<! (storage/delete-garbage-item conn index)))))
+
+(defn clean-garbage
+  "Cleans up garbage data for old indexes, but retains
+  the last `max-indexes` indexes. Note that if not retaining
+  at least one prior index, current queries or handles on the latest
+  db might fail, therefore `max-indexes` is recommended to typically
+  be 1 or more."
+  [{:keys [conn commit] :as _db} max-indexes]
+  (go-try
+   (assert (and (int? max-indexes)
+                (>= max-indexes 0)))
```

Comment:
If a `max-indexes` of zero will break queries some percentage of the time, should we just disallow it entirely?

Reply:
If we could validate this wherever it comes from instead of all the way down here where it's used I would feel better.

Reply:
> If a `max-indexes` of zero will break queries some percentage of the time, should we just disallow it entirely?

It will only break queries if you are holding onto an old `db` object which points to the old indexes - but that risk exists even if you have 100 old indexes. It is wholly dependent on how you use the API and how often index updates are happening - but in most circumstances it wouldn't present an issue at all.

Sometimes garbage can consume a lot of disk space, and if you are running this in a constrained environment and know the `0` setting won't impact how you use fluree, then this will be useful to have.

Reply:
> If we could validate this wherever it comes from instead of all the way down here where it's used I would feel better.

It mostly comes from the index-root that we write to disk, which is where we read the value from. We don't currently validate any index-related data on disk.

Reply:
Adjusted docstring to more accurately reflect risks of index garbage collection here: b8059a06afb624e55fd8282c4c8436a4df111f11